java/src/com/swiftnav/sbp/client/SBPFramer.java

```diff
@@ -0,0 +1,89 @@
+package com.swiftnav.sbp.client;
+
+import com.swiftnav.sbp.SBPBinaryException;
+import com.swiftnav.sbp.SBPMessage;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+public class SBPFramer extends SBPIterable implements SBPSender {
+    private static final byte PREAMBLE = 0x55;
+
+    private static final int PREAMBLE_SIZE = 1;
+    private static final int HEADER_SIZE = 5;
+    private static final int CRC_SIZE = 2;
+
+    private SBPDriver driver;
+
+    public SBPFramer(SBPDriver driver_) {
+        driver = driver_;
+    }
+
+    @Override
+    protected SBPMessage getNext() {
+        try {
+            return receive();
+        } catch (IOException e) {
+            throw new NoSuchElementException();
+        }
+    }
+
+    private SBPMessage receive() throws IOException {
+        /* Wait for a preamble to be received */
+        byte[] preamble;
+        do {
+            preamble = driver.read(PREAMBLE_SIZE);
+            if (preamble == null) {
+                throw new IOException("SBPDriver read() returned null!");
+            }
+        } while (preamble[0] != PREAMBLE);
+
+        /* Read the header */
+        byte[] header = driver.read(HEADER_SIZE);
+        int calccrc = CRC16.crc16(header);
+        /* Unpack fields from received header */
+        ByteBuffer bb = ByteBuffer.wrap(header).order(ByteOrder.LITTLE_ENDIAN);
+        int type = bb.getShort() & 0xffff;
+        int sender = bb.getShort() & 0xffff;
+        int len = bb.get() & 0xff;
+
+        byte[] payload = driver.read(len);
+        calccrc = CRC16.crc16(payload, calccrc);
+
+        int crc = ByteBuffer.wrap(driver.read(CRC_SIZE))
+                .order(ByteOrder.LITTLE_ENDIAN).getShort() & 0xffff;
+        if (crc != calccrc) {
+            System.err.println("CRC error in received SBPMessage");
```

Comment:
I don't know of a convenient way to do that.  They are not passed up in the C or Python clients.